### PR TITLE
Show images by tag on index page.

### LINKED ISSUE #4: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by lcguang**
Closes #4

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -20,7 +20,15 @@ def show
...
   def index
-    @images = Image.order(created_at: :desc)
+    if params[:tag].nil?
+      @images = Image.order(created_at: :desc)
+    else
+      @images = Image.tagged_with(params[:tag]).order(created_at: :desc)
+      if @images.empty?
+        flash[:warning] = "The tag #{params[:tag]} does not exist."
+        @images = Image.order(created_at: :desc)
+      end
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: This is good. But I also want to introduce the thin controller idea is that we want to keep controller as thin as possible and extract business logic out of controller because 1) we want controller just purely serve the responsibility of handling request and response, then test controller could be just simply test request and response 2) extract business logic out of controller could allow you unit test the logic. So I recommend to put these code to a `service` like https://github.com/AppFolioOnboarding/image-sharer-joshuasbates/blob/384b5979aa551ade8261ea7e00de24b1146bdc49/app/services/images_for_tag_name.rb.

Comment (lcguang) [reply app/controllers/images_controller.rb]: Thanks for the recommendation. We separated out that business logic into a `ImagesByTagService` and added tests for it. However, it seemed like adding this service should have reduced the amount of code in the controller tests, but we couldn't find any controller test pieces to remove. Do you have any suggestions?

Comment (shiboying) [reply app/controllers/images_controller.rb]: Hi, yes I think now in the controller test you should test 1) `tag_has_images` == true 2) `tag_has_images` != true. You don't need to test the nil tag case
